sway-lib-core/src/ops.sw

```diff
@@ -550,10 +485,186 @@ impl OrdEq for u8 {
+            w3 = shifted_3.add(carry_4);
+            w4 = shifted_4;
+        } else if w.eq(1) {
+            let(shifted_3, carry_3) = lsh_with_carry(word_3, b);
+            w1 = word_2.lsh(b).add(carry_3);
+            let(shifted_4, carry_4) = lsh_with_carry(word_4, b);
+            w2 = shifted_3.add(carry_4);
+            w3 = shifted_4;
+        } else if w.eq(2) {
+            let(shifted_4, carry_4) = lsh_with_carry(word_4, b);
+            w1 = word_3.lsh(b).add(carry_4);
+            w2 = shifted_4;
+        } else if w.eq(3) {
+            w1 = word_4.lsh(b);
+        } else {
+            ();
+        };
+
+        compose(w1, w2, w3, w4)
+    }
+
+    fn rsh(self, shift_amount: u64) -> Self {
+        let(word_1, word_2, word_3, word_4) = decompose(self);
+        let mut w1 = 0;
+        let mut w2 = 0;
+        let mut w3 = 0;
+        let mut w4 = 0;
+
+        let w = shift_amount.divide(64); // num of whole words to shift in addition to b
+        let b = shift_amount.modulo(64); // num of bits to shift within each word
+
+        // TODO: Use generalized looping version when vec lands !
+        if w.eq(0) {
+            let(shifted_3, carry_3) = rsh_with_carry(word_3, b);
+            w4 = word_4.rsh(b).add(carry_3);
+            let(shifted_2, carry_2) = rsh_with_carry(word_2, b);
+            w3 = shifted_3.add(carry_2);
+            let(shifted_1, carry_1) = rsh_with_carry(word_1, b);
+            w2 = shifted_2.add(carry_1);
+            w1 = shifted_1;
+        } else if w.eq(1) {
+            let(shifted_2, carry_2) = rsh_with_carry(word_2, b);
+            w4 = word_4.rsh(b).add(carry_2);
+            let(shifted_1, carry_1) = rsh_with_carry(word_1, b);
+            w3 = shifted_2.add(carry_1);
+            w2 = shifted_1;
+        } else if w.eq(2) {
+            let(shifted_1, carry_1) = rsh_with_carry(word_1, b);
+            w4 = word_2.rsh(b).add(carry_1);
+            w3 = shifted_1;
+        } else if w.eq(3) {
+            w4 = word_1.rsh(b);
+        } else {
+            ();
+        };
+
+        compose(w1, w2, w3, w4)
+    }
+}
+
 /////////////////////////////////////////////////
 // Internal Helpers
 /////////////////////////////////////////////////
 
+/// For setting the bit which allows overflow to occur without a vm panic
+const FLAG = 2;
+
+/// Left shift a u64 and preserve the overflow amount if any
+fn lsh_with_carry(word: u64, shift_amount: u64) -> (u64, u64) {
+    let mut output = (0, 0);
+    let right_shift_amount = 64.subtract(shift_amount);
+    let(shifted, carry) = asm(out: output, r1: word, r2: shift_amount, r3, r4, r5: right_shift_amount) {
+        srl r3 r1 r5; // shift right to get carry, put result in r3
+        sll r4 r1 r2; // shift left, put result in r4
+        sw out r4 i0; // store word at r4 in output
+        sw out r3 i1; // store word at r3 in output + 1 word offset
+        out: (u64, u64) // return both values
```

Comment:
Do this outside of asm. See: https://github.com/FuelLabs/sway/blob/609bcb42ccdba9d4a89aa1085e0fb3698ca2c41a/sway-lib-std/src/u128.sw#L195-L199